lib/ohloh_scm/git_scm.rb

```diff
@@ -2,5 +2,63 @@
 
 module OhlohScm
   class GitScm < Scm
+    def initialize(base:, url:, branch_name:, username:, password:)
+      super
+      @branch_name = branch_name || 'master'
+    end
+
+    # == Example:
+    #   remote_base = OhlohScm::Factory.get_base(url: 'https://github.com/ruby/ruby')
+    #   local_base = OhlohScm::Factory.get_base(url: '/tmp/ruby-src')
+    #   local_base.scm.pull(remote_base.scm)
+    def pull(from)
+      clone_or_fetch(from)
+    end
+
+    private
+
+    def clone_or_fetch(remote_scm)
+      if status.branch?(branch_name)
+        clean_and_checkout # should already be on correct branch, but we want to be careful.
+        run "cd '#{url}' && git fetch --tags --update-head-ok \
+              '#{remote_scm.url}' #{branch_name}:#{branch_name}"
+      else
+        clone_and_create_tracking_branch(remote_scm)
+      end
+      clean_up_disk
+    end
+
+    def clone_and_create_tracking_branch(remote_scm)
+      run "rm -rf '#{url}'"
```

Comment:
I noticed that we are no longer creating the directory to clone the source code into.  [run "mkdir -p '#{self.url}'"] . Is the directory being created somewhere else?  Is it expected to already exist, so the client should be creating the directory?

Reply:
Good find. I intended to explain this in the commit message, but I didn't! If we look at this piece of code in old ohloh_scm, this is what happens:
```
mkdir -p url
rm -rf url
git clone https://github.com/ruby/ruby url
```
Here `git clone` creates a new directory.
`Step 1` seems to have been added to avoid an error in `step 2`. However of **rm -rf**  does not fail when there is no directory. So `step 3` would execute regardless of `step 1`. Maybe `step 1` was a precautionary measure for older **rm -rf** implementations, however it doesn't seem applicable right now.

Reply:
I agree